x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/converters/normalize_rule_source.ts

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { RuleSource } from '../../../../../../../common/api/detection_engine';
+import { convertObjectKeysToSnakeCase } from '../../../../../../utils/object_case_converters';
+import type { BaseRuleParams } from '../../../../rule_schema';
+
+interface NormalizeRuleSourceParams {
+  immutable: BaseRuleParams['immutable'];
+  ruleSource: BaseRuleParams['ruleSource'];
+}
+export const normalizeRuleSource = ({
+  immutable,
+  ruleSource,
+}: NormalizeRuleSourceParams): RuleSource => {
+  if (!ruleSource) {
+    const normalizedRuleSource = immutable
+      ? {
+          type: 'external',
+          isCustomized: false,
+        }
+      : {
+          type: 'internal',
+        };
+
+    return convertObjectKeysToSnakeCase(normalizedRuleSource) as RuleSource;
+  }
+  return convertObjectKeysToSnakeCase(ruleSource) as RuleSource;
+};
```

Comment:
@xcrzx I remember you mentioning handling possible data inconsistencies. For example, data in ES being for whatever reason:
```
{
    "immutable": false,
    "ruleSource": {
        "type": "external",
        "isCustomized": true
    }
}
```
It's hard to me to think where these inconsitencies might arise from, but do you think it makes sense to rely always on `immutable` to calculate `rule_source`?

In the case above, calculating `rule_source` to be:
```
{
        "type": "internal"
}
```

Or if in ES data looks like:
```
{
    "immutable": true,
    "ruleSource": {
        "type": "internal",
    }
}
```
calculating `rule_source` to be:
```
{
        "type": "external",
        "isCustomized": false
}
```
WDYT?

Reply:
I think what you've implemented is correct. Relying on `immutable` in responses is not always possible because we also need to return `isCustomized`, which is calculated on writes. If the `ruleSource` field is available, we can return it as is. There's synchronization logic implemented on write, so `ruleSource` should always match `immutable` if present.